README.md
# Jsoner

Serialize HTML tables into JSON in Ruby.  
[![Build Status](https://travis-ci.org/simlegate/jsoner.png?branch=master)](https://travis-ci.org/simlegate/jsoner)

## Installation

Add this line to your application's Gemfile:

    gem 'jsoner'

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install jsoner

## Usage

```ruby
html = <<-eohtml
    <table id='example-table'>
      <thead>
        <tr>
          <th>First Name</th>
          <th>Last Name</th>
          <th>Points</th></tr>
      </thead>
      <tbody>
        <tr>
          <td>Jill</td>
          <td>Smith</td>
          <td>50</td></tr>
        <tr>
          <td>Eve</td>
          <td>Jackson</td>
          <td>94</td></tr>
        <tr>
          <td>John</td>
          <td>Doe</td>
          <td>80</td></tr>
        <tr>
          <td>Adam</td>
          <td>Johnson</td>
          <td>67</td></tr>
      </tbody>
    </table>
  eohtml

# Convert HTML table into Json

json = Jsoner.parse(html)

# output json => 
# 
#   [ {"First Name"=>"Jill", "Last Name"=>"Smith",   "Points"=>"50"},
#     {"First Name"=>"Eve",  "Last Name"=>"Jackson", "Points"=>"94"},
#     {"First Name"=>"John", "Last Name"=>"Doe",     "Points"=>"80"},
#     {"First Name"=>"Adam", "Last Name"=>"Johnson", "Points"=>"67"} ]

```
Or, `Jsoner` can parse HTML file including HTML table

```ruby
require 'open-uri'
# you must have table.html file and assign file path correctly.

table = Jsoner.parse(open('table.html'))

# output json => 
# 
#   [ {"First Name"=>"Jill", "Last Name"=>"Smith",   "Points"=>"50"},
#     {"First Name"=>"Eve",  "Last Name"=>"Jackson", "Points"=>"94"},
#     {"First Name"=>"John", "Last Name"=>"Doe",     "Points"=>"80"},
#     {"First Name"=>"Adam", "Last Name"=>"Johnson", "Points"=>"67"} ]

```

Or you can open Link

```ruby
require 'open-uri'

table = Jsoner.parse(open('http://www.w3school.com.cn/tiy/t.asp?f=html_table_header'))

# output json => 
#
# [{"姓名"=>"Bill Gates", "电话"=>"555 77 855"}]
```
## THANKS

[table-to-json](https://github.com/lightswitch05/table-to-json) written by [@lightswitch05](https://github.com/lightswitch05) in JavaScript.

## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request


CHANGELOG.md
## v0.0.2
* Filter other elements beyond table element
* Get first table element when having more than two in parsed HTML

## v0.0.3
* Parse HTML file including HTML Table
* Fixing file is empty or have no table element

## v0.0.4
* Add rspec test of checking integrity of HTML table
* Supply exception message of NotFullTable
* Parse Link including HTML Table

## v0.1.0
* Support Ruby 2.0.0


spec/jsoner_spec.rb
# encoding: utf-8
Dir["#{File.dirname(__FILE__)}/fixtures/*.rb"].each {|file| require file }
require "#{File.dirname(__FILE__)}/../lib/jsoner"

require 'open-uri'
require 'json'

describe Jsoner do
  
  it "should match data from json of fixtures" do
    Jsoner.parse(table_str).should == json
  end

  it "should parse file including table" do
    Jsoner.parse(open("#{File.dirname(__FILE__)}/fixtures/table.html")).should == json
  end

  it "should raise error when having no full table in HTML file" do
    expect{ Jsoner.parse(open("#{File.dirname(__FILE__)}/fixtures/table_extend.html"))}.to raise_error Jsoner::NotFullTable
  end

  # http://www.w3school.com.cn/tiy/t.asp?f=html_table_headers
  it "should parse Link include table" do
    json = Jsoner.parse(open("http://www.w3school.com.cn/tiy/t.asp?f=html_table_headers"))
    JSON.parse(json).should == [{"姓名"=>"Bill Gates", "电话"=>"555 77 855"}]
  end

  context "data parsed" do
    it "it is HTML file" do
      Jsoner.filter(open("#{File.dirname(__FILE__)}/fixtures/table.html")).should == table_str
    end

    it "it is  String" do
      Jsoner.filter("<table></table>").should == "<table></table>"
    end
  end
end


spec/fixtures/json.rb
require 'json'
def pre_json
  [ {"First Name"=>"Jill", "Last Name"=>"Smith",   "Points"=>"50"},
    {"First Name"=>"Eve",  "Last Name"=>"Jackson", "Points"=>"94"},
    {"First Name"=>"John", "Last Name"=>"Doe",     "Points"=>"80"},
    {"First Name"=>"Adam", "Last Name"=>"Johnson", "Points"=>"67"} ]
end

def json
  pre_json.to_json
end


spec/fixtures/table.rb
def table_str
  <<-eohtml
<table id='example-table'>
  <thead>
    <tr>
      <th>First Name</th>
      <th>Last Name</th>
      <th data-override="Score">Points</th></tr>
  </thead>
  <tbody>
    <tr>
      <td>Jill</td>
      <td>Smith</td>
      <td data-override="disqualified">50</td></tr>
    <tr>
      <td>Eve</td>
      <td>Jackson</td>
      <td>94</td></tr>
    <tr>
      <td>John</td>
      <td>Doe</td>
      <td>80</td></tr>
    <tr>
      <td>Adam</td>
      <td>Johnson</td>
      <td>67</td></tr>
  </tbody>
</table>
  eohtml
end


lib/jsoner/table.rb
require 'json'

module Jsoner
  class Table

    def initialize factory
      @table = factory.create
    end

    def row_number
      @table[:body].count
    end

    def to_json
      convert.to_json
    end

    #
    # convert Hash from factory into anthor Hash will be serialized into JSON
    # 
    # Example:
    #   table = { :header => ["First Name", "Last Name", "Points"]
    #             :body   => [["Jill", "Smith", "50"],
    #                         ["Eve", "Jackson", "94"],
    #                         ["John", "Doe", "80"],
    #                         ["Adam", "Johnson", "67"]] }
    #
    # Output:
    # table == [{"First Name"=>"Jill", "Last Name"=>"Smith",   "Points"=>"50"},
    #           {"First Name"=>"Eve",  "Last Name"=>"Jackson", "Points"=>"94"},
    #           {"First Name"=>"John", "Last Name"=>"Doe",     "Points"=>"80"},
    #           {"First Name"=>"Adam", "Last Name"=>"Johnson", "Points"=>"67"}]
    def convert
      (0...row_number).map do |index|

	#
	# Combine two Arrays into Hash
	#  a = ["a", "b", "c"]
	#  b = ["d", "e", "f"]
	#  Hash[a.zip(b)]
	#  => {"a" => "d", "b" => "e", "b" => "f" }
        Hash[@table[:header].zip(@table[:body][index])]
      end
    end

    private :convert
  end
end


lib/jsoner/table_factory.rb
module Jsoner
  #
  # build Hash below from doc parsed by Nokogiki
  #   table = { :header => ["First Name", "Last Name", "Score"]
  #             :body   => [["Jill", "Smith", "50"],
  #                         ["Eve", "Jackson", "94"],
  #                         ["John", "Doe", "80"],
  #                         ["Adam", "Johnson", "67"]
  #                        ]
  #           }
  class TableFactory

    def initialize doc
      @table_rows = remove_housing(doc).search('tr')
    end

    def create
      { :header => build_header, :body => build_body }
    end

    def build_header
      @table_rows[0].search('th').map(&:content)
    end

    def build_body
      row_number = @table_rows.count - 1
      (1..row_number).map do |row|
        @table_rows[row].search('td').map(&:content)
      end
    end

    # 
    # check whether table is full
    # a full table should include +th+, +td+, +tr+ elements
    def self.check doc
      table = doc.search('table').first
      unless table.search('tr').empty? || table.search('td').empty? || table.search('th').empty?
        TableFactory.new doc
      else
	raise Jsoner::NotFullTable 
      end
    end

    private
      
      #
      # Remove other elements beside table and keep it clean
      # Or keep one table when including more than two table elements in parsed HTML
      #
      # Example: 
      # <body>
      #   <table>
      #     // other elements
      #   </table>
      #   <table>
      #     // other elements
      #   </table>
      # </body>
      #
      # Output:
      #   <table>
      #     // other elements
      #   </table>
      def remove_housing doc
	doc.search('table').first
      end
  end 
end
